Add point and range indexes to future index section (#1273)
neo4j/neo4j-documentation: cypher/cypher-docs/src/docs/dev/ql/administration/indexes-for-search-performance/index.asciidoc

> Index provider can be specified using the `OPTIONS` clause.

Author: Add something about b-tree deprecated here as well, similar to how was done for drop by schema

[[administration-indexes-examples]]
== Creating indexes

> // TODO: add deprecated flag

Author: how do we do this here?

Reviewer: We have done it for some other tests like `Drop a single-property index`, but I cannot really find how it works

Author: Do we want to have the deprecated flag on all the examples or only the ones with btree keyword or options? (as that is what actually throws the warning)

Reviewer: I think it is enough with those that throw the warning, since we will probably reuse the other commands for range in 5.0

Author: Lets see if @emmaholmbergohlsson and @burqen agree :)

Reviewer 2: I agree with only the ones that will throw a warning. My guess for 5.0 was that we would change the default type and the syntax without explicit index type will still be allowed

Author: yes, then I'll remove the todos on the examples we don't want the flag on

(waiting for help on how to add the flag, Jessica is looking into it)

include::../indexes/drop-a-non-existing-index.asciidoc[leveloffset=+1]

> // TODO: should this section live here? or at some other level in the documentation

Author: To be discussed

Reviewer: I think the current placement, after all other index docs but before the deprecated parts is OK!

Author: I think most of us at least thinks it fits here, so lets keep it :)

include::../indexes/drop-a-non-existing-index.asciidoc[leveloffset=+1]

// TODO: should this section live here? or at some other level in the documentation
[[indexes-future-indexes]]
== Future indexes

> There are preparations for introducing two new types of indexes, point and range indexes.

Reviewer 2: Should we say something here about how usable they are and when they are coming or should this section just say you can read more about it here `<link-to-ops-manual>`?

Author: If we want to have examples of using the indexes similar to the query tuning section here as well I think we should add something but if not I think a link is sufficient

(I guess planner team would be responsible for such examples)

Reviewer 2: I was mostly thinking that we should mention that they aren't used yet :P

Reviewer: Yes I also think it make sense to add a sentence in the beginning about that they aren't used yet

Author: Will do (and then we update with link to operations manual when that exists)

Reviewer 3: ```suggestion
Two new types of indexes, point and range indexes, will be introduced in a future release.
```

These together with text indexes will replace the current b-tree indexes.

Like b-tree indexes, range indexes is created on one or more properties for all nodes or relationships with a given label or relationship type:

* An index that is created on a single property for any given label or relationship type is called a _single-property index_.
* An index that is created on more than one property for any given label or relationship type is called a _composite index_.

The differences in the usage patterns between composite and single-property indexes described in <<administration-indexes-single-vs-composite-index>> also applies to range indexes.

> Point indexes, like text indexes, are a kind of single-property indexes, with the limitation that they only recognize properties point values.

Reviewer 2: 'properties point values', is this missing a word?

Nodes or relationships with the indexed label or relationship type where the indexed property is of another value type are not included in the index.

=== Syntax

.Syntax for managing future indexes
[options="header", width="100%", cols="5a,3, 3a"]
|===
| Command | Description | Comment
| [source, cypher, role=noplay]
----
CREATE RANGE INDEX [index_name] [IF NOT EXISTS]
FOR (n:LabelName)
ON (n.propertyName_1[,
    n.propertyName_2,
    …
    n.propertyName_n])
[OPTIONS "{" option: value[, ...] "}"]
----
| Create a range index on nodes, either on a single property or composite.

Index provider can be specified using the `OPTIONS` clause.

.4+.^| Best practice is to give the index a name when it is created.
If the index is not explicitly named, it will get an auto-generated name.

The index name must be unique among both indexes and constraints.

The command is optionally idempotent, with the default behavior to throw an error if you attempt to create the same index twice.

> With `IF NOT EXISTS`, no error is thrown and nothing happens should an index with the same name, schema or both already exist.

Reviewer 2: 'the same name, schema or both' - type should probably be included here too?

Author: I seem to have missed this when I updated it in tests and for constraints at some point (and then copied it down here)

|===

=== Create index examples

include::../indexes/create-a-single-property-range-index-for-nodes.asciidoc[leveloffset=+1]

include::../indexes/create-a-single-property-range-index-for-relationships.asciidoc[leveloffset=+1]

include::../indexes/create-a-range-index-only-if-it-does-not-already-exist.asciidoc[leveloffset=+1]

include::../indexes/create-a-range-index-specifying-the-index-provider.asciidoc[leveloffset=+1]

include::../indexes/create-a-composite-range-index-for-nodes.asciidoc[leveloffset=+1]

include::../indexes/create-a-composite-range-index-for-relationships.asciidoc[leveloffset=+1]

include::../indexes/create-a-node-point-index.asciidoc[leveloffset=+1]

include::../indexes/create-a-relationship-point-index.asciidoc[leveloffset=+1]

include::../indexes/create-a-point-index-only-if-it-does-not-already-exist.asciidoc[leveloffset=+1]

include::../indexes/create-a-point-index-specifying-the-index-provider.asciidoc[leveloffset=+1]

include::../indexes/create-a-point-index-specifying-the-index-configuration.asciidoc[leveloffset=+1]

include::../indexes/create-a-point-index-specifying-both-the-index-provider-and-configuration.asciidoc[leveloffset=+1]

> 

Author: Also add examples for using the indexes here as its own sub-section, similar to the query-tuning chapter (probably a task for planner team)

Author: Decided to probably not do this

include::../indexes/drop-a-non-existing-index.asciidoc[leveloffset=+1]

[[indexes-future-indexes]]
== Future indexes

There are preparations for introducing two new types of indexes, point and range indexes.

> They cannot be used in queries yet, only created and dropped for migration purposes.

Reviewer 3: ```suggestion
They cannot be used in queries yet, but they can be created and dropped for migration purposes.
```

These new index types together with text indexes will replace the current b-tree indexes.
For more details on these new types, see the operations manual.
// TODO: add the link to the relevant operations manual section when that section exists

Like b-tree indexes, range indexes are created on one or more properties for all nodes or relationships with a given label or relationship type:

> * An index that is created on a single property for any given label or relationship type is called a _single-property index_.

Reviewer 3: ```suggestion
* An index created on a single property for any given label or relationship type is called a _single-property index_.
```

Author: this would be in 2 places then since I just copied this down here

> * An index that is created on more than one property for any given label or relationship type is called a _composite index_.

Reviewer 3: ```suggestion
* An index created on more than one property for any given label or relationship type is called a _composite index_.
```

The differences in the usage patterns between composite and single-property indexes described in <<administration-indexes-single-vs-composite-index>> also applies to range indexes.

Similar to b-tree indexes, range indexes may also back constraints by giving the range index provider when creating an index-backed constraint.

Point indexes, like text indexes, are a kind of single-property indexes, with the limitation that they only recognize properties with point values.
Nodes or relationships with the indexed label or relationship type where the indexed property is of another value type are not included in the index.

=== Syntax

.Syntax for managing future indexes
[options="header", width="100%", cols="5a,3, 3a"]
|===
| Command | Description | Comment
| [source, cypher, role=noplay]
----
CREATE RANGE INDEX [index_name] [IF NOT EXISTS]
FOR (n:LabelName)
ON (n.propertyName_1[,
    n.propertyName_2,
    …
    n.propertyName_n])
[OPTIONS "{" option: value[, ...] "}"]
----
| Create a range index on nodes, either on a single property or composite.

Index provider can be specified using the `OPTIONS` clause.

.4+.^| Best practice is to give the index a name when it is created.

> If the index is not explicitly named, it will get an auto-generated name.

Reviewer 3: ```suggestion
If the index is not explicitly named, it gets an auto-generated name.
```

Author: same here, multiple places

include::../indexes/drop-a-non-existing-index.asciidoc[leveloffset=+1]

[[indexes-future-indexes]]
== Future indexes

There are preparations for introducing two new types of indexes, point and range indexes.
They cannot be used in queries yet, only created and dropped for migration purposes.
These new index types together with text indexes will replace the current b-tree indexes.
For more details on these new types, see the operations manual.
// TODO: add the link to the relevant operations manual section when that section exists

> Like b-tree indexes, range indexes is created on one or more properties for all nodes or relationships with a given label or relationship type:

Reviewer 2: 'range indexes is created' should either be 'indexes are created' or 'index is created' I think?

Nodes or relationships with the indexed label or relationship type where the indexed property is of another value type are not included in the index.

=== Syntax

.Syntax for managing future indexes
[options="header", width="100%", cols="5a,3, 3a"]
|===
| Command | Description | Comment
| [source, cypher, role=noplay]
----
CREATE RANGE INDEX [index_name] [IF NOT EXISTS]
FOR (n:LabelName)
ON (n.propertyName_1[,
    n.propertyName_2,
    …
    n.propertyName_n])
[OPTIONS "{" option: value[, ...] "}"]
----
| Create a range index on nodes, either on a single property or composite.

Index provider can be specified using the `OPTIONS` clause.

.4+.^| Best practice is to give the index a name when it is created.
If the index is not explicitly named, it will get an auto-generated name.

The index name must be unique among both indexes and constraints.

The command is optionally idempotent, with the default behavior to throw an error if you attempt to create the same index twice.
With `IF NOT EXISTS`, no error is thrown and nothing happens should an index with the same name or same schema and index type already exist.

> It may still throw an error if conflicting constraints exist, such as constraints with the same name or schema.

Reviewer 2: Should 'constraints with the same name or schema.' also say 'schema and index type'?

Author: Probably
